junk-drawer/ecs.janet

```diff
@@ -1,92 +1,135 @@
+(import /junk-drawer/sparse-set :as sparse-set)
+(import /junk-drawer/cache :as cache)
+
+(def MAX_ENTITY_COUNT 1000)
+
 (defmacro def-component [name & fields]
-  "Define a new component with the specified fields."
-  (if (= 1 (length fields))
-    ~(defn ,name [value]
-       (assert
-        (= (type value) ,(first fields))
-        (string/format "%q must be of type %q" value ,(first fields)))
-       value)
-    (let [type-table (table ;fields)
-          def-array (mapcat |[$ (symbol $)] (keys type-table))]
-      ~(defn ,name [&keys ,(struct ;def-array)]
-         # assert types of the component fields
-         ,;(map
-            (fn [[key field-type]]
-              ~(assert
-                (= (type ,(symbol key)) ,field-type)
-                ,(string/format "%q must be of type %q" key field-type)))
-            (filter
-             |(not= (last $) :any)
-             (pairs type-table)))
-
-         # return the component
-         ,(table ;def-array)))))
+  "define a new component with the specified fields."
+  (let [type-table (table ;fields)
+        def-array (mapcat |[$ (symbol $)] (keys type-table))]
+    ~(defn ,name [&keys ,(struct ;def-array)]
+       # assert types of the component fields
+       ,;(map
+          (fn [[key field-type]]
+            ~(assert
+              (= (type ,(symbol key)) ,field-type)
+              ,(string/format "%q must be of type %q" key field-type)))
+          (filter
+           |(not= (last $) :any)
+           (pairs type-table)))
+
+       # return the component
+       ,(table ;def-array))))
 
 (defmacro def-tag [name]
-  "Define a new tag (component with no data)."
+  "define a new tag (component with no data)."
   ~(defn ,name [] true))
 
 (defmacro def-system [name queries & body]
-  "Define a system to do work on a list of queries."
+  "define a system to do work on a list of queries."
   ~(def ,name
      (tuple
        ,(values queries)
        (fn [,;(keys queries) dt] ,;body))))
 
-(defmacro add-entity [world & components]
-  "Add a new entity with the given components to the world."
-  (with-syms [$id $db $wld]
+(defmacro add-component [world eid component]
+  (with-syms [$wld $cmp-name]
     ~(let [,$wld ,world
-           ,$id (get ,$wld :id-counter)
-           ,$db (get ,$wld :database)]
-       (put-in ,$db [:entity ,$id] ,$id)
-       ,;(map
-           |(quasiquote (put-in ,$db [,(keyword (first $)) ,$id] ,$))
-           components)
-       (put ,$wld :id-counter (inc ,$id))
-       ,$id)))
+           ,$cmp-name ,(keyword (first component))]
+       (when (nil? (get-in ,$wld [:database ,$cmp-name]))
+         (put-in ,$wld
+                  [:database ,$cmp-name]
+                  (sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)))
```

Comment:
I would unquote the `sparse-set/init` function here, which could solve some of the code shuffling in this PR. It also applies to some more functions in the quasi quotes.

Reply:
I had a lot of problems with this macro, so I'll definitely try that out!

Reply:
It is one of the very good Janet  macro tricks